File: ui/android-app/app/src/main/java/org/aerobag/app/domain/SessionUpdateAccumulator.kt
// SPDX-FileCopyrightText: 2026 Aerobag contributors
//
// SPDX-License-Identifier: AGPL-3.0-or-later

package org.aerobag.app.domain

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.longOrNull
import org.aerobag.app.generated.UiSessionUpdate
import org.aerobag.app.generated.UiSessionUpdateGroup
import org.aerobag.app.generated.projectionPatches

internal enum class SessionUpdateDisposition {
    Applied,
    Stale,
    ResyncRequired,
}

internal data class SessionUpdateApplication(
    val disposition: SessionUpdateDisposition,
    val changedGroups: Set<UiSessionUpdateGroup> = emptySet(),
    val changedPaths: Set<List<String>> = emptySet(),
    val installedFullSnapshot: Boolean = false,
)

internal class SessionUpdateContractException(message: String) : IllegalArgumentException(message)

internal class SessionUpdateAccumulator(
    initialSnapshot: JsonObject,
    private val expectedContractVersion: Int,
    private val json: Json,
) {
    private val groupVersions = mutableMapOf<UiSessionUpdateGroup, Long>()

    var snapshot: JsonObject = sanitizeFullSnapshot(initialSnapshot)
        private set

    fun replaceFullSnapshot(value: JsonElement): JsonObject {
        val nextSnapshot = sanitizeFullSnapshot(value)
        if (nextSnapshot.wireLong("session_revision") < snapshot.wireLong("session_revision")) {
            return snapshot
        }
        snapshot = nextSnapshot
        groupVersions.clear()
        return snapshot
    }

    fun apply(value: JsonElement): SessionUpdateDisposition = applyDetailed(value).disposition

    fun applyDetailed(value: JsonElement): SessionUpdateApplication =
        applyDetailed(parseSessionUpdate(value))

    fun applyOrResync(
        value: JsonElement,
        loadFullSnapshot: () -> JsonElement,
    ): SessionUpdateDisposition = applyOrResyncDetailed(value, loadFullSnapshot).disposition

    fun applyOrResyncDetailed(
        value: JsonElement,
        loadFullSnapshot: () -> JsonElement,
    ): SessionUpdateApplication {
        val application = applyDetailed(value)
        if (application.disposition == SessionUpdateDisposition.ResyncRequired) {
            replaceFullSnapshot(loadFullSnapshot())
            return application.copy(installedFullSnapshot = true)
        }
        return application
    }

    private fun applyDetailed(update: UiSessionUpdate): SessionUpdateApplication {
        if (update.uiContractVersion != expectedContractVersion) {
            throw SessionUpdateContractException(
                "UI wire contract ${update.uiContractVersion} is unsupported; " +
                    "client requires $expectedContractVersion",
            )
        }
        val currentRevision = snapshot.wireLong("session_revision")
        if (update.sessionRevision <= currentRevision) {
            return SessionUpdateApplication(SessionUpdateDisposition.Stale)
        }
        if (update.sessionRevision != currentRevision + 1) {
            return SessionUpdateApplication(SessionUpdateDisposition.ResyncRequired)
        }

        var nextSnapshot: JsonElement = snapshot
        val assignedPaths = mutableListOf<List<String>>()
        val nextVersions = mutableListOf<Pair<UiSessionUpdateGroup, Long>>()
        val changedGroups = mutableSetOf<UiSessionUpdateGroup>()
        for ((group, patch) in update.projectionPatches()) {
            val previousVersion = groupVersions[group]
            if (previousVersion != null && patch.version <= previousVersion) {
                throw SessionUpdateContractException(
                    "session update ${group.wireName} version ${patch.version} " +
                        "does not advance $previousVersion",
                )
            }
            for ((index, assignment) in patch.assignments.withIndex()) {
                val envelopeField = assignment.path.first()
                if (envelopeField in ENVELOPE_FIELDS) {
                    throw SessionUpdateContractException(
                        "session update ${group.wireName} cannot replace envelope field $envelopeField",
                    )
                }
                val overlap = assignedPaths.firstOrNull { pathsOverlap(it, assignment.path) }
                if (overlap != null) {
                    throw SessionUpdateContractException(
                        "session update path ${assignment.path.joinToString("/")} " +
                            "overlaps ${overlap.joinToString("/")}",
                    )
                }
                nextSnapshot = replaceAtPath(
                    nextSnapshot,
                    assignment.path,
                    0,
                    assignment.value,
                    "session update ${group.wireName} assignment $index",
                )
                assignedPaths += assignment.path
            }
            nextVersions += group to patch.version
            changedGroups += group
        }

        snapshot = JsonObject(
            (nextSnapshot as? JsonObject
                ?: throw SessionUpdateContractException("session update result must be an object")) + mapOf(
                "ui_contract_version" to JsonPrimitive(update.uiContractVersion),
                "session_revision" to JsonPrimitive(update.sessionRevision),
            ),
        )
        nextVersions.forEach { (group, version) -> groupVersions[group] = version }
        return SessionUpdateApplication(
            disposition = SessionUpdateDisposition.Applied,
            changedGroups = changedGroups,
            changedPaths = assignedPaths.toSet(),
        )
    }

    private fun parseSessionUpdate(value: JsonElement): UiSessionUpdate {
        val update = value as? JsonObject
            ?: throw SessionUpdateContractException("session update must be a JSON object")
        rejectUnknownKeys(
            update,
            setOf("ui_contract_version", "session_revision") +
                UiSessionUpdateGroup.entries.map(UiSessionUpdateGroup::wireName),
            "session update",
        )
        update.wireLong("ui_contract_version")
        update.wireLong("session_revision")
        for (group in UiSessionUpdateGroup.entries) {
            val patch = update[group.wireName] ?: continue
            if (patch is JsonNull) continue
            val patchObject = patch as? JsonObject
                ?: throw SessionUpdateContractException(
                    "session update ${group.wireName} patch must be a JSON object",
                )
            rejectUnknownKeys(
                patchObject,
                setOf("version", "assignments"),
                "session update ${group.wireName} patch",
            )
            patchObject.wireLong("version")
            val assignments = patchObject["assignments"] as? JsonArray
            if (assignments == null) {
                throw SessionUpdateContractException(
                    "session update ${group.wireName} assignments must be an array",
                )
            }
            assignments.forEachIndexed { index, assignmentValue ->
                val label = "session update ${group.wireName} assignment $index"
                val assignment = assignmentValue as? JsonObject
                    ?: throw SessionUpdateContractException("$label must be an object")
                rejectUnknownKeys(assignment, setOf("path", "value"), label)
                val path = assignment["path"] as? JsonArray
                    ?: throw SessionUpdateContractException("$label path must be a nonempty array")
                if (path.isEmpty()) {
                    throw SessionUpdateContractException("$label path must be a nonempty array")
                }
                if (path.any { segment ->
                        segment !is JsonPrimitive || !segment.isString || segment.content.isEmpty()
                    }
                ) {
                    throw SessionUpdateContractException(
                        "$label path segments must be nonempty strings",
                    )
                }
                if ("value" !in assignment) {
                    throw SessionUpdateContractException("$label must contain value")
                }
            }
        }
        return json.decodeFromJsonElement(UiSessionUpdate.serializer(), update)
    }

    private fun pathsOverlap(left: List<String>, right: List<String>): Boolean =
        (0 until minOf(left.size, right.size)).all { index -> left[index] == right[index] }

    private fun replaceAtPath(
        current: JsonElement,
        path: List<String>,
        offset: Int,
        value: JsonElement,
        label: String,
    ): JsonElement {
        if (offset == path.size) return value
        val segment = path[offset]
        return when (current) {
            is JsonObject -> {
                val child = current[segment]
                    ?: throw SessionUpdateContractException(
                        "$label path does not exist at $segment",
                    )
                JsonObject(current + (segment to replaceAtPath(child, path, offset + 1, value, label)))
            }
            is JsonArray -> {
                val index = segment.toIntOrNull()
                    ?.takeIf { it >= 0 && it.toString() == segment }
                    ?: throw SessionUpdateContractException(
                        "$label path segment $segment is not an array index",
                    )
                if (index !in current.indices) {
                    throw SessionUpdateContractException("$label array index $index is out of range")
                }
                JsonArray(current.mapIndexed { childIndex, child ->
                    if (childIndex == index) {
                        replaceAtPath(child, path, offset + 1, value, label)
                    } else {
                        child
                    }
                })
            }
            else -> throw SessionUpdateContractException("$label path parent must be an object or array")
        }
    }

    private fun sanitizeFullSnapshot(value: JsonElement): JsonObject {
        val raw = value as? JsonObject
            ?: throw SessionUpdateContractException("session snapshot must be a JSON object")
        if ("session_update" in raw) {
            throw SessionUpdateContractException(
                "full session snapshot must not contain session_update",
            )
        }
        val snapshot = raw
        val contractVersion = snapshot.wireLong("ui_contract_version")
        snapshot.wireLong("session_revision")
        if (contractVersion != expectedContractVersion.toLong()) {
            throw SessionUpdateContractException(
                "UI wire contract $contractVersion is unsupported; client requires $expectedContractVersion",
            )
        }
        return snapshot
    }

    private fun JsonObject.wireLong(field: String): Long {
        val primitive = get(field) as? JsonPrimitive
        val value = primitive
            ?.takeUnless { it.isString }
            ?.longOrNull
            ?: throw SessionUpdateContractException("$field must be a non-negative integer")
        if (value < 0) throw SessionUpdateContractException("$field must be a non-negative integer")
        return value
    }

    private fun rejectUnknownKeys(value: JsonObject, allowed: Set<String>, label: String) {
        val unknown = value.keys - allowed
        if (unknown.isNotEmpty()) {
            throw SessionUpdateContractException("$label has unknown fields: ${unknown.joinToString()}")
        }
    }

    private companion object {
        val ENVELOPE_FIELDS = setOf("ui_contract_version", "session_revision", "session_update")
    }
}

File: ui/web-app/src/domain/sessionUpdateAccumulator.ts
// SPDX-FileCopyrightText: 2026 Aerobag contributors
//
// SPDX-License-Identifier: AGPL-3.0-or-later

import {
  UI_SESSION_UPDATE_GROUPS,
  type UiSessionProjectionPatch,
  type UiSessionProjectionAssignment,
  type UiSessionUpdate,
  type UiSessionUpdateGroup,
} from "../generated/sessionUpdateWire";

type JsonObject = Record<string, unknown>;

export type SessionUpdateDisposition = "applied" | "stale" | "resync_required";
export type SessionProjectionDisposition = SessionUpdateDisposition | "full_snapshot";

export class SessionUpdateContractError extends Error {}

function isJsonObject(value: unknown): value is JsonObject {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function requireJsonObject(value: unknown, label: string): JsonObject {
  if (!isJsonObject(value)) {
    throw new SessionUpdateContractError(`${label} must be a JSON object`);
  }
  return value;
}

function requireWireInteger(value: unknown, label: string): number {
  if (!Number.isSafeInteger(value) || (value as number) < 0) {
    throw new SessionUpdateContractError(`${label} must be a non-negative safe integer`);
  }
  return value as number;
}

function rejectUnknownKeys(value: JsonObject, allowed: ReadonlySet<string>, label: string): void {
  const unknown = Object.keys(value).filter((key) => !allowed.has(key));
  if (unknown.length > 0) {
    throw new SessionUpdateContractError(`${label} has unknown fields: ${unknown.join(", ")}`);
  }
}

function parsePatch(value: unknown, group: UiSessionUpdateGroup): UiSessionProjectionPatch | null {
  if (value === undefined || value === null) return null;
  const patch = requireJsonObject(value, `session update ${group} patch`);
  rejectUnknownKeys(patch, new Set(["version", "assignments"]), `session update ${group} patch`);
  requireWireInteger(patch.version, `session update ${group} version`);
  if (!Array.isArray(patch.assignments)) {
    throw new SessionUpdateContractError(`session update ${group} assignments must be an array`);
  }
  for (const [index, value] of patch.assignments.entries()) {
    parseAssignment(value, `session update ${group} assignment ${index}`);
  }
  return patch as unknown as UiSessionProjectionPatch;
}

function parseAssignment(value: unknown, label: string): UiSessionProjectionAssignment {
  const assignment = requireJsonObject(value, label);
  rejectUnknownKeys(assignment, new Set(["path", "value"]), label);
  if (!Array.isArray(assignment.path) || assignment.path.length === 0) {
    throw new SessionUpdateContractError(`${label} path must be a nonempty array`);
  }
  for (const segment of assignment.path) {
    if (typeof segment !== "string" || segment.length === 0) {
      throw new SessionUpdateContractError(`${label} path segments must be nonempty strings`);
    }
  }
  if (!Object.hasOwn(assignment, "value")) {
    throw new SessionUpdateContractError(`${label} must contain value`);
  }
  return assignment as unknown as UiSessionProjectionAssignment;
}

function pathsOverlap(left: readonly string[], right: readonly string[]): boolean {
  const commonLength = Math.min(left.length, right.length);
  for (let index = 0; index < commonLength; index += 1) {
    if (left[index] !== right[index]) return false;
  }
  return true;
}

function replaceAtPath(
  current: unknown,
  path: readonly string[],
  offset: number,
  value: unknown,
  label: string,
): unknown {
  if (offset === path.length) return value;
  const segment = path[offset];
  if (Array.isArray(current)) {
    if (!/^(0|[1-9][0-9]*)$/.test(segment)) {
      throw new SessionUpdateContractError(`${label} path segment ${segment} is not an array index`);
    }
    const index = Number(segment);
    if (index >= current.length) {
      throw new SessionUpdateContractError(`${label} array index ${index} is out of range`);
    }
    const next = [...current];
    next[index] = replaceAtPath(current[index], path, offset + 1, value, label);
    return next;
  }
  const object = requireJsonObject(current, `${label} path parent`);
  if (!Object.hasOwn(object, segment)) {
    throw new SessionUpdateContractError(`${label} path does not exist at ${segment}`);
  }
  return {
    ...object,
    [segment]: replaceAtPath(object[segment], path, offset + 1, value, label),
  };
}

export function parseSessionUpdate(value: unknown): UiSessionUpdate {
  const update = requireJsonObject(value, "session update");
  rejectUnknownKeys(
    update,
    new Set(["ui_contract_version", "session_revision", ...UI_SESSION_UPDATE_GROUPS]),
    "session update",
  );
  requireWireInteger(update.ui_contract_version, "session update contract version");
  requireWireInteger(update.session_revision, "session update revision");
  for (const group of UI_SESSION_UPDATE_GROUPS) {
    parsePatch(update[group], group);
  }
  return update as unknown as UiSessionUpdate;
}

function sanitizedFullSnapshot(value: unknown, expectedContractVersion: number): JsonObject {
  const raw = requireJsonObject(value, "session snapshot");
  if (Object.hasOwn(raw, "session_update")) {
    throw new SessionUpdateContractError("full session snapshot must not contain session_update");
  }
  const snapshot = { ...raw };
  const contractVersion = requireWireInteger(snapshot.ui_contract_version, "snapshot contract version");
  requireWireInteger(snapshot.session_revision, "snapshot revision");
  if (contractVersion !== expectedContractVersion) {
    throw new SessionUpdateContractError(
      `UI wire contract ${contractVersion} is unsupported; client requires ${expectedContractVersion}`,
    );
  }
  return snapshot;
}

export class SessionUpdateAccumulator {
  private rawSnapshot: JsonObject;
  private readonly groupVersions = new Map<UiSessionUpdateGroup, number>();

  constructor(initialSnapshot: unknown, private readonly expectedContractVersion: number) {
    this.rawSnapshot = sanitizedFullSnapshot(initialSnapshot, expectedContractVersion);
  }

  get snapshot(): JsonObject {
    return this.rawSnapshot;
  }

  replaceFullSnapshot(value: unknown): JsonObject {
    const nextSnapshot = sanitizedFullSnapshot(value, this.expectedContractVersion);
    if ((nextSnapshot.session_revision as number) < (this.rawSnapshot.session_revision as number)) {
      return this.rawSnapshot;
    }
    this.rawSnapshot = nextSnapshot;
    this.groupVersions.clear();
    return this.rawSnapshot;
  }

  apply(value: unknown): SessionUpdateDisposition {
    const update = parseSessionUpdate(value);
    if (update.ui_contract_version !== this.expectedContractVersion) {
      throw new SessionUpdateContractError(
        `UI wire contract ${update.ui_contract_version} is unsupported; client requires ${this.expectedContractVersion}`,
      );
    }
    const currentRevision = this.rawSnapshot.session_revision as number;
    if (update.session_revision <= currentRevision) return "stale";
    if (update.session_revision !== currentRevision + 1) return "resync_required";

    let nextSnapshot: JsonObject = this.rawSnapshot;
    const assignedPaths: Array<readonly string[]> = [];
    const nextVersions: Array<[UiSessionUpdateGroup, number]> = [];
    for (const group of UI_SESSION_UPDATE_GROUPS) {
      const patch = parsePatch(update[group], group);
      if (!patch) continue;
      const previousVersion = this.groupVersions.get(group);
      if (previousVersion !== undefined && patch.version <= previousVersion) {
        throw new SessionUpdateContractError(
          `session update ${group} version ${patch.version} does not advance ${previousVersion}`,
        );
      }
      for (const [index, assignmentValue] of patch.assignments.entries()) {
        const assignment = parseAssignment(
          assignmentValue,
          `session update ${group} assignment ${index}`,
        );
        const envelopeField = assignment.path[0];
        if (envelopeField === "ui_contract_version" || envelopeField === "session_revision" || envelopeField === "session_update") {
          throw new SessionUpdateContractError(`session update ${group} cannot replace envelope field ${envelopeField}`);
        }
        const overlap = assignedPaths.find((path) => pathsOverlap(path, assignment.path));
        if (overlap) {
          throw new SessionUpdateContractError(
            `session update path ${assignment.path.join("/")} overlaps ${overlap.join("/")}`,
          );
        }
        nextSnapshot = requireJsonObject(
          replaceAtPath(
            nextSnapshot,
            assignment.path,
            0,
            assignment.value,
            `session update ${group}`,
          ),
          `session update ${group} result`,
        );
        assignedPaths.push(assignment.path);
      }
      nextVersions.push([group, patch.version]);
    }

    this.rawSnapshot = {
      ...nextSnapshot,
      ui_contract_version: update.ui_contract_version,
      session_revision: update.session_revision,
    };
    for (const [group, version] of nextVersions) this.groupVersions.set(group, version);
    return "applied";
  }

  async applyOrResync(
    value: unknown,
    loadFullSnapshot: () => Promise<unknown>,
  ): Promise<SessionUpdateDisposition> {
    const disposition = this.apply(value);
    if (disposition === "resync_required") {
      this.replaceFullSnapshot(await loadFullSnapshot());
    }
    return disposition;
  }

  async applyProjectionResult(
    value: unknown,
    loadFullSnapshot: () => Promise<unknown>,
  ): Promise<SessionProjectionDisposition> {
    if (isJsonObject(value) && Object.hasOwn(value, "app_ui_state")) {
      this.replaceFullSnapshot(value);
      return "full_snapshot";
    }
    return this.applyOrResync(value, loadFullSnapshot);
  }

}
